api/agent/pure_runner.go

```diff
@@ -864,6 +920,29 @@ func (pr *pureRunner) Status(ctx context.Context, _ *empty.Empty) (*runner.Runne
 	return pr.handleStatusCall(ctx)
 }
 
+// BeforeCall called before a function is executed
+func (pr *pureRunner) BeforeCall(ctx context.Context, call *models.Call) error {
+	if call.Type != models.TypeDetached {
+		return nil
+	}
+	var err error
+	// it is an ack sync we send ResultStart message back
+	pr.callHandleLock.Lock()
+	ch := pr.callHandleMap[call.ID]
```

Comment:
I don't think we can put the channel in the context as we can't pass it via gRPC protocol. This side of the code is the pure runner so leaving on a different machine and to communicate via the runner_client it needs to go via gRPC, the communication can't go via channel, am I right?
We could decide to change the runner protocol to allow a special message to signal the ack but we have discarded this solution at the beginning when we started to reason about this change. I think @skinowski suggested to not mess with the runner protocol as it is not necessary if we use the call listener.

Reply:
@reclaro In this case, you'd be adding the channel inside the ctx in the runner already. So, it does not need to cross gRPC.

Reply:
hold on I am confused now, which channel are we talking about?  the only channel I have is on the LB side to return from the Submit method and this channel can't cross gRPC.
Are here we talking about putting the callHandler in the context? is the misunderstanding coming from the name of the variable called ch? if so then yes I agree that the callHandler is just a thing for runner and agent so it is not crossing the gRPC pond. I am always a bit reluctant to abuse the context but as Reed said the callHandler is thread local so it could be ok to add it here.

Reply:
mmh I looked at it, the pure_runner is where we create the callHandler then the pure_runner calls the agent.Submit passing just the call. Adding the callhandler to the context is not easy, we can pollute the call.req.Context() adding the callHandler or adding a new context field to the call just to store the callHandler? I don't think the latter is a good solution at that point I'd add a callHandler field to the call but that even more dirty as the callHandler has a field for the call so we have a kind of loop dependency which seems wrong. The other solution is to leave the lock in place here, I heard your voice as regards lock nightmare, is there anything we can do to make it more solid?

Reply:
that's okay, it was just an idea anyway. this code appears to work and we can improve later.